javatests/arcs/android/storage/database/DatabaseImplTest.kt

```diff
@@ -1218,6 +1218,112 @@ class DatabaseImplTest {
+            schema,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+        val entity2 = DatabaseData.Entity(
+            RawEntity(
+                "entity2", 
+                mapOf("text" to "def".toReferencable()), 
+                mapOf("nums" to setOf(123.0.toReferencable(), 789.0.toReferencable())),
+                11L,
+                timeInPast // expirationTimestamp, in the past.
+            ),
+            schema,
+            FIRST_VERSION_NUMBER,
+            VERSION_MAP
+        )
+        // Add both of them to a collection.
+        val values = setOf(
+            Reference("entity1", backingKey, VersionMap("ref1" to 1)),
+            Reference("entity2", backingKey, VersionMap("ref2" to 2))
+        )
+        val collection = DatabaseData.Collection(
+            values = values,
+            schema = schema,
+            databaseVersion = FIRST_VERSION_NUMBER,
+            versionMap = VERSION_MAP
+        )
+
+        database.insertOrUpdate(entity1Key, entity1)
+        database.insertOrUpdate(entity2Key, entity2)
+        database.insertOrUpdate(collectionKey, collection)
+
+        database.removeExpiredEntities()
+
+        // Check the expired entities fields have been cleared (only a tombstone is left).
+        assertThat(database.getEntity(entity1Key, schema))
+            .isEqualTo(DatabaseData.Entity(
+                RawEntity(
+                    "entity1", 
+                    mapOf("text" to null), 
+                    mapOf("nums" to emptySet()),
+                    11L,
+                    timeInPast
+                ),
+                schema,
+                FIRST_VERSION_NUMBER,
+                VERSION_MAP
+            ))
+        assertThat(database.getEntity(entity2Key, schema))
+            .isEqualTo(DatabaseData.Entity(
+                RawEntity(
+                    "entity2", 
+                    mapOf("text" to null), 
+                    mapOf("nums" to emptySet()),
+                    11L,
+                    timeInPast
+                ),
+                schema,
+                FIRST_VERSION_NUMBER,
+                VERSION_MAP
+            ))
+
+        // Check the collection is empty.
+        assertThat(database.getCollection(collectionKey, schema))
+            .isEqualTo(collection.copy(values = setOf()))
+        
+        // Check unused values have been deleted from the global table as well.
+        assertTableIsEmpty("field_values")
+
+        // Check collection entries have been cleared.
+        assertTableIsEmpty("collection_entries")
+
+        // Check the collections for nums are gone (the collection left is the entity collection).
+        assertTableIsSize("collections", 1)
+
+        // Check the entity refs are gone.
+        assertTableIsEmpty("entity_refs")
+
+        // Check unused primitive values have been removed.
+        assertTableIsEmpty("text_primitive_values")
+        assertTableIsEmpty("number_primitive_values")
```

Comment:
Optional: this method could accept varargs, that would make these assertions a bit shorter. Not really necessary though.

Reply:
Having the asserts on two different lines has the benefit that they fail separately. Ie if you see the test failing on the assert at line 1323 you know the problem is in text_primitive_values. I think a method checking more than one table at a time would make debugging failures less easy.

Reply:
You can always use `assertWithMessage`:

```
fun assertEmptyTables(vararg tables: String) {
    tables.forEach { tableName -> {
        val count = getTableSize(tableName)
        assertWithMessage("Expected $tableName to be empty but had $count rows.")
            .that(count)
            .isEqualTo(0)
```

Reply:
yes, but my point was mostly about the stack strace: the stack trace will look nicer because you can see more easily which of the two tables failed.
What is the advantage of using the vararg function? you are just shortening the test code, not making it any nicer I think.

Reply:
I am merging this as is for now, but let me know if you think I really should change this and I'll do a follow up.

Reply:
What you have is fine too, all good.